compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java

```diff
@@ -8185,4 +8166,295 @@ private boolean isEndOfTypeList(SyntaxKind nextTokenKind) {
+        return parseTableConstructorOrQueryRhs(peek().kind, tableKeyword, keySpecifier);
+    }
+
+    private STNode parseTableConstructorOrQueryRhs(SyntaxKind nextTokenKind, STNode tableKeyword, STNode keySpecifier) {
+        switch (nextTokenKind) {
+            case FROM_KEYWORD:
+                return parseQueryExprRhs(parseQueryConstructType(tableKeyword, keySpecifier));
+            case OPEN_BRACKET_TOKEN:
+                return parseTableConstructorExprRhs(tableKeyword, keySpecifier);
+            default:
+                Solution solution = recover(peek(), ParserRuleContext.TABLE_CONSTRUCTOR_OR_QUERY_RHS,
+                    tableKeyword, keySpecifier);
+
+                // If the parser recovered by inserting a token, then try to re-parse the same
+                // rule with the inserted token. This is done to pick the correct branch
+                // to continue the parsing.
+                if (solution.action == Action.REMOVE) {
+                    return solution.recoveredNode;
+                }
+
+                return parseTableConstructorOrQueryRhs(solution.tokenKind, tableKeyword, keySpecifier);
+        }
+    }
+
+    /**
+     * Parse query expression.
+     * <code>query-expr-rhs := query-pipeline select-clause</code>
+     *
+     * @param queryConstructType queryConstructType that precedes this rhs
+     * @return Parsed node
+     */
+    private STNode parseQueryExprRhs(STNode queryConstructType) {
+        switchContext(ParserRuleContext.QUERY_EXPRESSION);
+        STNode queryPipeline = parseQueryPipeline();
+        STNode selectClause = parseSelectClause();;
+        endContext();
+        return STNodeFactory.createQueryExpressionNode(queryConstructType, queryPipeline, selectClause);
+    }
+
+    /**
+     * Parse query construct type.
+     * <p>
+     * <code>query-construct-type := table key-specifier</code>
+     *
+     * @return Parsed node
+     */
+    private STNode parseQueryConstructType(STNode tableKeyword, STNode keySpecifier) {
+        return STNodeFactory.createQueryConstructTypeNode(tableKeyword, keySpecifier);
+    }
+
+    /**
+     * Parse query pipeline.
+     * <p>
+     * <code>
+     * query-pipeline := from-clause intermediate-clause*
+     * <br/>
+     * intermediate-clause := from-clause | where-clause | let-clause
+     * </code>
+     *
+     * @return Parsed node
+     */
+    private STNode parseQueryPipeline() {
+        STNode fromClause = parseFromClause();
+        List<STNode> clauses = new ArrayList<>();
+        return parseQueryPipeline(fromClause, clauses);
+    }
+
+    private STNode parseQueryPipeline(STNode fromClause, List<STNode> clauses) {
+        return parseQueryPipeline(peek().kind, fromClause, clauses);
+    }
+    private STNode parseQueryPipeline(SyntaxKind nextTokenKind, STNode fromClause, List<STNode> clauses) {
+        STNode clause;
+
+        while (true) {
```

Comment:
This is a bit dangerous, as we may get stuck in an infinite loop. Better to do a `while(!isEndOfExpression())`